src/Microsoft.AspNetCore.Server.HttpSys/MessagePump.cs

```diff
@@ -250,28 +252,55 @@ private static void SetFatalResponse(RequestContext context, int status)
 
         public Task StopAsync(CancellationToken cancellationToken)
         {
-            _stopping = true;
-            // Wait for active requests to drain
-            if (_outstandingRequests > 0)
+            void RegisterCancelation()
             {
-                LogHelper.LogInfo(_logger, "Stopping, waiting for " + _outstandingRequests + " request(s) to drain.");
-
-                var waitForStop = new TaskCompletionSource<object>();
                 cancellationToken.Register(() =>
                 {
-                    LogHelper.LogInfo(_logger, "Timed out, terminating " + _outstandingRequests + " request(s).");
-                    waitForStop.TrySetResult(0);
+                    if (Interlocked.Exchange(ref _shutdownSignalCompleted, 1) == 0)
+                    {
+                        LogHelper.LogInfo(_logger, "Canceled, terminating " + _outstandingRequests + " request(s).");
+                        _shutdownSignal.TrySetResult(null);
+                    }
                 });
+            }
 
-                return Task.WhenAny(_shutdownSignal.Task, waitForStop.Task);
+            if (Interlocked.Exchange(ref _stopping, 1) == 1)
+            {
+                if (_outstandingRequests > 0)
```

Comment:
Why do you even check this? It's a race condition. Always call RegisterCancelation